Implement reference picker

julien-nc opened the pull request:
refs https://github.com/nextcloud/server/issues/31667
requires https://github.com/nextcloud/server/pull/35557

* [x] implement reference picker
    * [x] first step: reference provider picker (multi select)
    * [x] second step: vanilla input to enter a link
    * [x] alternative second step if provider is searchable: multi select to search with all search providers supported by the reference provider
* [x] add mechanism to allow apps to inject a custom picker component

The reference picker gets the reference provider list via an initial state provided by the server. Whatever page using the reference picker should have dispatched the `RenderReferenceEvent` to trigger the production of the initial state.

Each reference provider can declare a list of supported unified search providers. Those search provider will be used if no custom link picker is provided. The search UI is pretty similar to the core unified search: small thumbnail, a title and a sub line.

If the search UI does not fit the needs of the reference provider, it is possible to register a custom component to let users choose an element to get a link in the end. This component must emit the `submit` event with the link as event data. It an also emit the `cancel` event to go back to the reference provider picker.

To test this PR, one can:
* use https://github.com/nextcloud/server/pull/35557
* use the `reference-discovery` branch of the integration_github app (don't forget to link with `@nextcloud/vue-richtext` while having the `enh/noid/reference-picker` branch checked out)
* go to the "Connected accounts" admin settings where a test textarea+button can be found to use the reference picker

Remaining questions/issues/todos:
* There is no translations in vue-richtext but this brings strings that should be translated.
* We should find a way to use this reference picker in the `RichText` component (when typing a special character and/or when receiving an event via the NC event bus and/or via a component method call)

julien-nc commented:
Here is a little demo with 3 reference providers:
* basic "GitHub permalink" with no supported search provider but a custom picker element registered (a button to submit  a hardcoded result)
* "GitHub issue/pr" which supports 2 search providers
* Giphy (in development) which registers a complex custom picker element (a modal to have a nicer UI to search for gifs)

https://user-images.githubusercontent.com/11291457/208720450-c0f73f62-28be-4be6-800d-cd325dbfdec2.mp4

julien-nc commented:
@nimishavijay @jancborchardt Do you like it? :grin: Any obvious visible design mistake?

juliushaertl commented on src/referencePicker/CustomPickerElement.vue:
```diff
@@ -0,0 +1,67 @@
+			}
+
+			this.renderResult = renderCustomPickerElement(this.$refs.customElement, { providerId: this.provider.id, accessible: false })
+			if (this.renderResult.object.$on) {
```
```suggestion
			if (this.renderResult?.object?.$on) {
```

juliushaertl replied:
Maybe also check if object is a vue instance here

jancborchardt reviewed (changes requested):
Very cool! :) Some design feedback:
- [ ] Would the dropdown element on the right not open in place of where you type `#` (or `/`) like in Notion, instead of off to the right?
- [ ] The "GitHub code permalink" option is a bit strange since you can just put in any link and it will just be put in the text – you could just type it in the text itself or paste it. Is there a benefit doing it this way? Otherwise I’d say let’s remove the entry?
- [ ] The standard from Notion seems to be the `/` character to invoke these kinds of pickers. Could we switch to that? (We can also support both so people familiar with Notion and GitHub are familiar.)
- [x] The Giphy gif picker could use a nice emptycontent view. In messaging apps that is often just popular Gifs, Giphy’s API probaby has that?

juliushaertl commented:
Notes from the design call we had earlier today with @nimishavijay 

- Search on custom pickers would be nice (e.g. for the file picker)
- Could already be done by having two providers (pick file / search file)
- Refine what you are looking for
  - Possibly provide hints for generic search providers like `repo:...` syntax for github
- Probably nice for demo purposes:
  - YouTube provider with searach
- How to type for first input
  - Behaviour should keep the text typed in the input field like in notion/emoji picker
  - We probably should rethink about which character to pick to not interfere with markdown typing
- It is up to developers to be able to list their providers even if there is no fancy search
  - Mainly to increase visibility of the feature

jancborchardt commented:
@julien-nc added a point to my review:

> The standard from Notion seems to be the `/` character to invoke these kinds of pickers. Could we switch to that? (We can also support both so people familiar with Notion and GitHub are familiar.)

julien-nc commented:
@jancborchardt Thanks for the review. Triggering the link picker with `/` might conflict with Talk commands. Let's talk about that in the design review call.

The overall appearance has changed since the screencast. Choosing the link provider can now be done like choosing an emoji.

julien-nc commented:
https://user-images.githubusercontent.com/11291457/213427705-774b0c7d-2c0a-4f8c-8fa0-4f504f2d210d.mp4

Current state:
* 3 types of providers
    * "raw": Just there to let users know a certain type of link will be rendered. It lets users enter a link and shows a preview on the fly
    * "search": Linked with one or more search providers. It lets users enter a search query and shows combined results from all related search providers
    * "custom picker": Shows a custom component registered by the reference provider

There are 3 ways to open the link picker:
* Just use the ReferencePicker component anywhere we want
* Using the ReferencePickerModal component (or the helper function "getLinkWithPicker" which takes care of everything and returns a promise which resolves with the link
* In the RichContentEditable component by typing "/" and selecting a provider. The picker will is opened with the provider already selected

jancborchardt reviewed (approved):
Wow, really amazing! :) I think it would be best to merge it so we can see it in action in different environments like Text, Talk, comments etc.

Only detail note: When writing "GitHub issues, pull reuquests […]", the "p" of pull requests is capital, but should be lowercase. :)

juliushaertl commented on src/referencePicker/RawLinkInput.vue:
```diff
@@ -0,0 +1,162 @@
+			loading: false,
+			reference: null,
+			abortController: null,
+			// TODO translate?
```
Let's have this as a follow up task so we can get a first release out.

julien-nc replied:
Ok and let's keep the comments if you don't mind.

juliushaertl commented on src/referencePicker/RawLinkInput.vue:
```diff
@@ -0,0 +1,162 @@
+			return isUrl(this.inputValue)
+		},
+	},
+	mounted() {
+	},
+	beforeDestroy() {
+	},
```
🔥

juliushaertl commented on src/referencePicker/ReferencePickerModal.vue:
```diff
@@ -0,0 +1,170 @@
+import ArrowLeftIcon from 'vue-material-design-icons/ArrowLeft.vue'
+import CloseIcon from 'vue-material-design-icons/Close.vue'
+
+import NcModal from '@nextcloud/vue/dist/Components/NcModal.js'
+import NcButton from '@nextcloud/vue/dist/Components/NcButton.js'
```
```suggestion
import { NcModal, NcButton } from '@nextcloud/vue'
```

Let's stick to the generic import in all places since this will make sure that `@nextcloud/vue` is not bundled within the vue-richtext package but apps can pull this in.

julien-nc replied:
NcHighlight needs to be directly imported, all the other ones are imported with the generic import now.

juliushaertl reviewed (approved):
Just minor comments, otherwise good to go 🚀

julien-nc commented on src/referencePicker/CustomPickerElement.vue:
```diff
@@ -0,0 +1,65 @@
+			renderResult: null,
+		}
+	},
+	computed: {
```
Sure

juliushaertl commented:
Sounds good

julien-nc commented:
@juliushaertl Rebased on master, added JS debug log for providers without related search providers or without custom picker component.
Cleaned up a little bit. Ready for review.
